src/fmeval/eval_algorithms/faithfulness.py

```diff
@@ -0,0 +1,297 @@
+context:\n$context
+statements:\n$statements
+Answer: Assistant:
+"""  # noqa: E501
+
+
+class FaithfulnessScore(Transform):
+    """This transform augments its input record with the computed faithfulness score.
+
+    See the docstring for `Faithfulness` for more details regarding the score itself.
+    """
+
+    def __init__(
+        self,
+        output_key: str = FAITHFULNESS,
+    ):
+        """FaithfulnessScore initializer.
+
+        :param output_key: The key corresponding to the faithfulness score that
+            will be added to the input record.
+        """
+        super().__init__(output_key)
+        self.register_input_output_keys(
+            input_keys=[],
+            output_keys=[output_key],
+        )
+        self.output_key = output_key
+
+    @validate_call
+    def __call__(self, record: Dict[str, Any]) -> Dict[str, Any]:
+        """Augment the input record with the computed faithfulness score.
+
+        :param record: The input record.
+        :returns: The input record, with the faithfulness score added in.
+        """
+        verdict_output = record["raw_verdicts"]
+        statements = record["statements"]
+        record[self.output_key] = self._get_score(verdict_output, statements)
+        return record
+
+    @staticmethod
+    def _get_score(verdict_output: str, statements: str) -> float:
+        """Given generated statements and verdicts, compute Faithfulness score.
+
+        :param verdict_output: Verdicts(Yes/No) and explanations string get from Judge model.
+        :param statements: Statements string get from `GetStatements` Transform.
+        :returns: 0 to 1. See the docstring for `Faithfulness` for more details
+            on what these numerical values represent.
+        """
+        output = verdict_output.lower().strip()
+        num_statements = len(statements.split("\n"))
+        # TODO: handle edge case that num_statements is 0?
+
+        # calculate score
+        final_answer = "Final verdicts in order:"
+        final_answer = final_answer.lower()
+        if output.find(final_answer) != -1:
+            output = output[output.find(final_answer) + len(final_answer) :]
+            score = float(sum(1 if "yes" in answer else 0 for answer in output.strip().split(".") if answer != ""))
+            score = score / num_statements
+        else:
+            score = float(max(0, output.count("verdict: yes")) / num_statements)
+
+        return score
+
+
+class GetStatements(Transform):
+    """This transform invokes the judge model string from raw statements."""
+
+    def __init__(self, input_key: str, output_key: str, judge_model: ModelRunner):
+        """GetStatements Initializer.
+        :param input_key: The key corresponding to prompt to get statements.
+        :param output_key: The key corresponding to the statements, which gets added to the record.
+        :param judge_model: An instance of ModelRunner representing the judge model to be used.
```

Comment:
Based on this doc string, it looks like `judge_model` should be Optional. Like so: `judge_model: Optional[ModelRunner] = None`. But, the code doesn't look like it is doing that? Should the doc string be updated?

Reply:
`judge_model` is required during these intermediate transform steps. It's optional if we provide a default, which is at the time when calling `evaluate` and `evaluate_sample`.